### PR TITLE
Factor out common DiagnosingTestCase support.

### PR DESCRIPTION
**Pull request by allevato**
Over time we've built up three slightly different versions of the
same `DiagnostingTestCase` class in different test targets. This
change factors out all the commonalities into a separate test
support target and has the other tests now subclass it.

On platforms that support it (macOS 10.15+, and Linux), formatting
tests now use `CollectionDifference` to show a line-based diff of
the expected/actual output, which is a significant readability
improvement. For example:

```
failed - Pretty-printed result was not what was expected - Actual output (+) differed from expected output (-):
 func name<R>(
   _ x: Int
 ) throws -> R
-where
-  Fooooooo == Barrrrr
+  where
+    Fooooooo == Barrrrr
```

### CODE REVIEW

**Comment by allevato** [comment]
cc @dylansturg

**Comment by allevato** [comment]
To clarify from the original description, when it says "formatting tests  now use `CollectionDifference`", that's both pretty printer tests *and* phase 1 rule tests—everything goes through the same comparison logic now.

File: Tests/SwiftFormatWhitespaceLinterTests/WhitespaceTestCase.swift [hunk 1]
```diff
@@ -1,54 +1,23 @@
...
-  }
-
-  override func tearDown() {
-    guard shouldCheckForUnassertedDiagnostics else { return }
-
-    for diag in consumer.registeredDiagnostics {
-      XCTFail("unexpected diagnostic '\(diag)' emitted")
-    }
+class WhitespaceTestCase: DiagnosingTestCase {
+  override func setUp() {
```
Comment (dylansturg) [inline Tests/SwiftFormatWhitespaceLinterTests/WhitespaceTestCase.swift]: Should this call `super.setUp`? Any chance of `DiagnosingTestCase.setUp` containing some code that needs to be executed?

Comment (allevato) [reply Tests/SwiftFormatWhitespaceLinterTests/WhitespaceTestCase.swift]: Yes! Even though all we do here is change the value of `shouldCheckForUnassertedDiagnostics`, if we ever add more logic in the base class we'll want to make sure it's not skipped.

File: Sources/SwiftFormatTestSupport/DiagnosingTestCase.swift [hunk 2]
```diff
@@ -0,0 +1,170 @@
+import SwiftFormatConfiguration
+import SwiftFormatCore
+import SwiftSyntax
+import XCTest
+
+/// DiagnosingTestCase is an XCTestCase subclass meant to inject diagnostic-specific testing
+/// routines into specific formatting test cases.
+open class DiagnosingTestCase: XCTestCase {
+  /// Set during lint tests to indicate that we should check for any unasserted diagnostics when the
```
[the remark this answers is not in the document]
Comment (allevato) [reply Sources/SwiftFormatTestSupport/DiagnosingTestCase.swift]: Done.

**Pull request merged**